rnp fails to build on GNU/Hurd (`PATH_MAX` is not available)

Over on the debian Hurd build daemon, [rnp fails to build from source](https://buildd.debian.org/status/fetch.php?pkg=rnp&arch=hurd-i386&ver=0.15.2-4&stamp=1635606066&raw=0).  (Hurd [has no PATH_MAX](https://www.gnu.org/software/hurd/hurd/porting/guidelines.html#PATH_MAX_tt_MAX_PATH_tt_MAXPATHL))

Here's some compiler errors stumbling on the problem:

```
[  4%] Building CXX object src/lib/CMakeFiles/librnp-obj.dir/__/librepgp/stream-common.cpp.o
cd /<<PKGBUILDDIR>>/build/src/lib && /usr/bin/c++ -D_GNU_SOURCE -Dlibrnp_EXPORTS -I/<<PKGBUILDDIR>>/build/src/lib -I/<<PKGBUILDDIR>>/src/common -I/<<PKGBUILDDIR>>/include -I/<<PKGBUILDDIR>>/src/lib -I/<<PKGBUILDDIR>>/src -isystem /usr/include/botan-2 -isystem /usr/include/json-c -g -O2 -ffile-prefix-map=/<<PKGBUILDDIR>>=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -fPIC -fvisibility=hidden -Wall -Wextra -Wunreachable-code -Wpointer-arith -Wmissing-declarations -Wno-pedantic -Wno-ignored-qualifiers -Wno-unused-parameter -Wno-missing-field-initializers -std=c++11 -MD -MT src/lib/CMakeFiles/librnp-obj.dir/__/librepgp/stream-common.cpp.o -MF CMakeFiles/librnp-obj.dir/__/librepgp/stream-common.cpp.o.d -o CMakeFiles/librnp-obj.dir/__/librepgp/stream-common.cpp.o -c /<<PKGBUILDDIR>>/src/librepgp/stream-common.cpp
/<<PKGBUILDDIR>>/src/librepgp/stream-common.cpp:728:15: error: ‘PATH_MAX’ was not declared in this scope
  728 |     char path[PATH_MAX];
      |               ^~~~~~~~
/<<PKGBUILDDIR>>/src/librepgp/stream-common.cpp: In function ‘void file_dst_close(pgp_dest_t*, bool)’:
/<<PKGBUILDDIR>>/src/librepgp/stream-common.cpp:766:31: error: ‘pgp_dest_file_param_t’ {aka ‘struct pgp_dest_file_param_t’} has no member named ‘path’
  766 |             rnp_unlink(param->path);
      |                               ^~~~
/<<PKGBUILDDIR>>/src/librepgp/stream-common.cpp: In function ‘rnp_result_t init_fd_dest(pgp_dest_t*, int, const char*)’:
/<<PKGBUILDDIR>>/src/librepgp/stream-common.cpp:779:35: error: ‘pgp_dest_file_param_t’ {aka ‘struct pgp_dest_file_param_t’} has no member named ‘path’
  779 |     if (path_len >= sizeof(param->path)) {
      |                                   ^~~~
/<<PKGBUILDDIR>>/src/librepgp/stream-common.cpp:789:19: error: ‘pgp_dest_file_param_t’ {aka ‘struct pgp_dest_file_param_t’} has no member named ‘path’
  789 |     memcpy(param->path, path, path_len + 1);
      |                   ^~~~
/<<PKGBUILDDIR>>/src/librepgp/stream-common.cpp: In function ‘rnp_result_t init_file_dest(pgp_dest_t*, const char*, bool)’:
/<<PKGBUILDDIR>>/src/librepgp/stream-common.cpp:806:35: error: ‘pgp_dest_file_param_t’ {aka ‘struct pgp_dest_file_param_t’} has no member named ‘path’
  806 |     if (path_len >= sizeof(param->path)) {
      |                                   ^~~~
/<<PKGBUILDDIR>>/src/librepgp/stream-common.cpp: In function ‘rnp_result_t file_tmpdst_finish(pgp_dest_t*)’:
/<<PKGBUILDDIR>>/src/librepgp/stream-common.cpp:858:37: error: ‘PATH_MAX’ was not declared in this scope
  858 |     char                   origpath[PATH_MAX] = {0};
      |                                     ^~~~~~~~
/<<PKGBUILDDIR>>/src/librepgp/stream-common.cpp:865:27: error: ‘pgp_dest_file_param_t’ {aka ‘struct pgp_dest_file_param_t’} has no member named ‘path’
  865 |     plen = strnlen(param->path, sizeof(param->path));
      |                           ^~~~
/<<PKGBUILDDIR>>/src/librepgp/stream-common.cpp:865:47: error: ‘pgp_dest_file_param_t’ {aka ‘struct pgp_dest_file_param_t’} has no member named ‘path’
  865 |     plen = strnlen(param->path, sizeof(param->path));
      |                                               ^~~~
/<<PKGBUILDDIR>>/src/librepgp/stream-common.cpp:869:13: error: ‘origpath’ was not declared in this scope
  869 |     strncpy(origpath, param->path, plen - strlen(TMPDST_SUFFIX));
      |             ^~~~~~~~
/<<PKGBUILDDIR>>/src/librepgp/stream-common.cpp:869:30: error: ‘pgp_dest_file_param_t’ {aka ‘struct pgp_dest_file_param_t’} has no member named ‘path’
  869 |     strncpy(origpath, param->path, plen - strlen(TMPDST_SUFFIX));
      |                              ^~~~
/<<PKGBUILDDIR>>/src/librepgp/stream-common.cpp:895:27: error: ‘pgp_dest_file_param_t’ {aka ‘struct pgp_dest_file_param_t’} has no member named ‘path’
  895 |     if (rnp_rename(param->path, origpath)) {
      |                           ^~~~
/<<PKGBUILDDIR>>/src/librepgp/stream-common.cpp: In function ‘void file_tmpdst_close(pgp_dest_t*, bool)’:
/<<PKGBUILDDIR>>/src/librepgp/stream-common.cpp:916:31: error: ‘pgp_dest_file_param_t’ {aka ‘struct pgp_dest_file_param_t’} has no member named ‘path’
  916 |             rnp_unlink(param->path);
      |                               ^~~~
/<<PKGBUILDDIR>>/src/librepgp/stream-common.cpp: In function ‘rnp_result_t init_tmpfile_dest(pgp_dest_t*, const char*, bool)’:
/<<PKGBUILDDIR>>/src/librepgp/stream-common.cpp:927:32: error: ‘PATH_MAX’ was not declared in this scope
  927 |     char                   tmp[PATH_MAX];
      |                                ^~~~~~~~
/<<PKGBUILDDIR>>/src/librepgp/stream-common.cpp:932:21: error: ‘tmp’ was not declared in this scope; did you mean ‘tm’?
  932 |     ires = snprintf(tmp, sizeof(tmp), "%s%s", path, TMPDST_SUFFIX);
      |                     ^~~
      |                     tm
src/lib/CMakeFiles/librnp-obj.dir/build.make:92: recipe for target 'src/lib/CMakeFiles/librnp-obj.dir/__/librepgp/stream-common.cpp.o' failed
make[3]: *** [src/lib/CMakeFiles/librnp-obj.dir/__/librepgp/stream-common.cpp.o] Error 1
```

--- a/src/common/str-utils.cpp
+++ b/src/common/str-utils.cpp
@@ -53,6 +53,20 @@ strip_eol(char *s)
 }
 
 bool
+strip_eol(std::string &s)
+{
+    size_t len = s.size();
+    while (len && ((s[len - 1] == '\n') || (s[len - 1] == '\r'))) {
+        len--;
+    }
+    if (len == s.size()) {
+        return false;
+    }
+    s.resize(len);
+    return true;
+}
+
+bool
 is_blank_line(const char *line, size_t len)
 {
     for (size_t i = 0; i < len && line[i]; i++) {

--- a/src/common/str-utils.h
+++ b/src/common/str-utils.h
@@ -27,10 +27,19 @@
 #ifndef RNP_STR_UTILS_H_
 #define RNP_STR_UTILS_H_
 
+#include <string>
+
 namespace rnp {
 char *strip_eol(char *s);
-bool  is_blank_line(const char *line, size_t len);
-bool  str_case_eq(const char *s1, const char *s2);
+/**
+ * @brief Strip EOL characters from the string's end.
+ *
+ * @param s string to check
+ * @return true if EOL was found and stripped, or false otherwise.
+ */
+bool strip_eol(std::string &s);
+bool is_blank_line(const char *line, size_t len);
+bool str_case_eq(const char *s1, const char *s2);
 } // namespace rnp
 #ifdef _WIN32
 #include <string>

--- a/src/librepgp/stream-common.cpp
+++ b/src/librepgp/stream-common.cpp
@@ -49,6 +49,7 @@
 #include "file-utils.h"
 #include "crypto/mem.h"
 #include <algorithm>
+#include <memory>
 
 bool
 src_read(pgp_source_t *src, void *buf, size_t len, size_t *readres)
@@ -725,16 +726,15 @@ dst_close(pgp_dest_t *dst, bool discard)
 }
 
 typedef struct pgp_dest_file_param_t {
-    int  fd;
-    int  errcode;
-    bool overwrite;
-    char path[PATH_MAX];
+    int         fd;
+    int         errcode;
+    bool        overwrite;
+    std::string path;
 } pgp_dest_file_param_t;
 
 static rnp_result_t
 file_dst_write(pgp_dest_t *dst, const void *buf, size_t len)
 {
-    ssize_t                ret;
     pgp_dest_file_param_t *param = (pgp_dest_file_param_t *) dst->param;
 
     if (!param) {
@@ -743,7 +743,7 @@ file_dst_write(pgp_dest_t *dst, const void *buf, size_t len)
     }
 
     /* we assyme that blocking I/O is used so everything is written or error received */
-    ret = write(param->fd, buf, len);
+    ssize_t ret = write(param->fd, buf, len);
     if (ret < 0) {
         param->errcode = errno;
         RNP_LOG("write failed, error %d", param->errcode);
@@ -758,7 +758,6 @@ static void
 file_dst_close(pgp_dest_t *dst, bool discard)
 {
     pgp_dest_file_param_t *param = (pgp_dest_file_param_t *) dst->param;
-
     if (!param) {
         return;
     }
@@ -766,52 +765,42 @@ file_dst_close(pgp_dest_t *dst, bool discard)
     if (dst->type == PGP_STREAM_FILE) {
         close(param->fd);
         if (discard) {
-            rnp_unlink(param->path);
+            rnp_unlink(param->path.c_str());
         }
     }
 
-    free(param);
+    delete param;
     dst->param = NULL;
 }
 
 static rnp_result_t
 init_fd_dest(pgp_dest_t *dst, int fd, const char *path)
 {
-    pgp_dest_file_param_t *param;
-    size_t                 path_len = strlen(path);
-    if (path_len >= sizeof(param->path)) {
-        RNP_LOG("path too long");
-        return RNP_ERROR_BAD_PARAMETERS;
-    }
-    if (!init_dst_common(dst, sizeof(*param))) {
+    if (!init_dst_common(dst, 0)) {
         return RNP_ERROR_OUT_OF_MEMORY;
     }
 
-    param = (pgp_dest_file_param_t *) dst->param;
-    param->fd = fd;
-    memcpy(param->path, path, path_len + 1);
+    try {
+        std::unique_ptr<pgp_dest_file_param_t> param(new pgp_dest_file_param_t());
+        param->path = path;
+        param->fd = fd;
+        dst->param = param.release();
+    } catch (const std::exception &e) {
+        RNP_LOG("%s", e.what());
+        return RNP_ERROR_OUT_OF_MEMORY;
+    }
+
     dst->write = file_dst_write;
     dst->close = file_dst_close;
     dst->type = PGP_STREAM_FILE;
-
     return RNP_SUCCESS;
 }
 
 rnp_result_t
 init_file_dest(pgp_dest_t *dst, const char *path, bool overwrite)
 {
-    int                    fd;
-    int                    flags;
-    struct stat            st;
-    pgp_dest_file_param_t *param;
-
-    size_t path_len = strlen(path);
-    if (path_len >= sizeof(param->path)) {
-        RNP_LOG("path too long");
-        return RNP_ERROR_BAD_PARAMETERS;
-    }
-
     /* check whether file/dir already exists */
+    struct stat st;
     if (!rnp_stat(path, &st)) {
         if (!overwrite) {
             RNP_LOG("file already exists: '%s'", path);
@@ -827,7 +816,7 @@ init_file_dest(pgp_dest_t *dst, const char *path, bool overwrite)
         }
     }
 
-    flags = O_WRONLY | O_CREAT;
+    int flags = O_WRONLY | O_CREAT;
     flags |= overwrite ? O_TRUNC : O_EXCL;
 #ifdef HAVE_O_BINARY
     flags |= O_BINARY;
@@ -836,8 +825,7 @@ init_file_dest(pgp_dest_t *dst, const char *path, bool overwrite)
     flags |= _O_BINARY;
 #endif
 #endif
-    fd = rnp_open(path, flags, S_IRUSR | S_IWUSR);
-
+    int fd = rnp_open(path, flags, S_IRUSR | S_IWUSR);
     if (fd < 0) {
         RNP_LOG("failed to create file '%s'. Error %d.", path, errno);
         return RNP_ERROR_WRITE;
@@ -856,58 +844,57 @@ static rnp_result_t
 file_tmpdst_finish(pgp_dest_t *dst)
 {
     pgp_dest_file_param_t *param = (pgp_dest_file_param_t *) dst->param;
-    size_t                 plen = 0;
-    struct stat            st;
-    char                   origpath[PATH_MAX] = {0};
-
     if (!param) {
         return RNP_ERROR_BAD_PARAMETERS;
     }
 
-    /* remove suffix so we have required path */
-    plen = strnlen(param->path, sizeof(param->path));
-    if (plen < strlen(TMPDST_SUFFIX)) {
-        return RNP_ERROR_BAD_PARAMETERS;
-    }
-    strncpy(origpath, param->path, plen - strlen(TMPDST_SUFFIX));
-
-    /* rename the temporary file */
+    /* close the file */
     close(param->fd);
     param->fd = -1;
 
-    /* check if file already exists */
-    if (!rnp_stat(origpath, &st)) {
-        if (!param->overwrite) {
-            RNP_LOG("target path already exists");
-            return RNP_ERROR_BAD_STATE;
-        }
+    /* rename the temporary file */
+    if (param->path.size() < strlen(TMPDST_SUFFIX)) {
+        return RNP_ERROR_BAD_PARAMETERS;
+    }
+    try {
+        /* remove suffix so we have required path */
+        std::string origpath(param->path.begin(), param->path.end() - strlen(TMPDST_SUFFIX));
+        /* check if file already exists */
+        struct stat st;
+        if (!rnp_stat(origpath.c_str(), &st)) {
+            if (!param->overwrite) {
+                RNP_LOG("target path already exists");
+                return RNP_ERROR_BAD_STATE;
+            }
 #ifdef _WIN32
-        /* rename() call on Windows fails if destination exists */
-        else {
-            rnp_unlink(origpath);
-        }
+            /* rename() call on Windows fails if destination exists */
+            else {
+                rnp_unlink(origpath.c_str());
+            }
 #endif
 
-        /* we should remove dir if overwriting, file will be unlinked in rename call */
-        if (S_ISDIR(st.st_mode) && rmdir(origpath)) {
-            RNP_LOG("failed to remove directory");
+            /* we should remove dir if overwriting, file will be unlinked in rename call */
+            if (S_ISDIR(st.st_mode) && rmdir(origpath.c_str())) {
+                RNP_LOG("failed to remove directory");
+                return RNP_ERROR_BAD_STATE;
+            }
+        }
+
+        if (rnp_rename(param->path.c_str(), origpath.c_str())) {
+            RNP_LOG("failed to rename temporary path to target file: %s", strerror(errno));
             return RNP_ERROR_BAD_STATE;
         }
-    }
-
-    if (rnp_rename(param->path, origpath)) {
-        RNP_LOG("failed to rename temporary path to target file: %s", strerror(errno));
+        return RNP_SUCCESS;
+    } catch (const std::exception &e) {
+        RNP_LOG("%s", e.what());
         return RNP_ERROR_BAD_STATE;
     }
-
-    return RNP_SUCCESS;
 }
 
 static void
 file_tmpdst_close(pgp_dest_t *dst, bool discard)
 {
     pgp_dest_file_param_t *param = (pgp_dest_file_param_t *) dst->param;
-
     if (!param) {
         return;
     }
@@ -916,43 +903,44 @@ file_tmpdst_close(pgp_dest_t *dst, bool discard)
     if (!dst->finished && (dst->type == PGP_STREAM_FILE)) {
         close(param->fd);
         if (discard) {
-            rnp_unlink(param->path);
+            rnp_unlink(param->path.c_str());
         }
     }
 
-    free(param);
+    delete param;
     dst->param = NULL;
 }
 
 rnp_result_t
 init_tmpfile_dest(pgp_dest_t *dst, const char *path, bool overwrite)
 {
-    char                   tmp[PATH_MAX];
-    pgp_dest_file_param_t *param = NULL;
-    rnp_result_t           res = RNP_ERROR_GENERIC;
-    int                    ires = 0;
-
-    ires = snprintf(tmp, sizeof(tmp), "%s%s", path, TMPDST_SUFFIX);
-    if ((ires < 0) || ((size_t) ires >= sizeof(tmp))) {
-        RNP_LOG("failed to build file path");
-        return RNP_ERROR_BAD_PARAMETERS;
-    }
+    try {
+        std::string tmp = std::string(path) + std::string(TMPDST_SUFFIX);
+        /* make sure tmp.data() is zero-terminated */
+        tmp.push_back('\0');
 #if defined(HAVE_MKSTEMP) && !defined(_WIN32)
-    int fd = mkstemp(tmp);
+        int fd = mkstemp(&tmp[0]);
 #else
-    int fd = rnp_mkstemp(tmp);
+        int fd = rnp_mkstemp(&tmp[0]);
 #endif
-    if (fd < 0) {
-        RNP_LOG("failed to create temporary file with template '%s'. Error %d.", tmp, errno);
-        return RNP_ERROR_WRITE;
-    }
-    if ((res = init_fd_dest(dst, fd, tmp))) {
-        close(fd);
-        return res;
+        if (fd < 0) {
+            RNP_LOG("failed to create temporary file with template '%s'. Error %d.",
+                    tmp.c_str(),
+                    errno);
+            return RNP_ERROR_WRITE;
+        }
+        rnp_result_t res = init_fd_dest(dst, fd, tmp.c_str());
+        if (res) {
+            close(fd);
+            return res;
+        }
+    } catch (const std::exception &e) {
+        RNP_LOG("%s", e.what());
+        return RNP_ERROR_BAD_STATE;
     }
 
     /* now let's change some parameters to handle temporary file correctly */
-    param = (pgp_dest_file_param_t *) dst->param;
+    pgp_dest_file_param_t *param = (pgp_dest_file_param_t *) dst->param;
     param->overwrite = overwrite;
     dst->finish = file_tmpdst_finish;
     dst->close = file_tmpdst_close;
@@ -962,18 +950,11 @@ init_tmpfile_dest(pgp_dest_t *dst, const char *path, bool overwrite)
 rnp_result_t
 init_stdout_dest(pgp_dest_t *dst)
 {
-    pgp_dest_file_param_t *param;
-
-    if (!init_dst_common(dst, sizeof(*param))) {
-        return RNP_ERROR_OUT_OF_MEMORY;
+    rnp_result_t res = init_fd_dest(dst, STDOUT_FILENO, "");
+    if (res) {
+        return res;
     }
-
-    param = (pgp_dest_file_param_t *) dst->param;
-    param->fd = STDOUT_FILENO;
-    dst->write = file_dst_write;
-    dst->close = file_dst_close;
     dst->type = PGP_STREAM_STDOUT;
-
     return RNP_SUCCESS;
 }
 

--- a/src/rnp/fficli.cpp
+++ b/src/rnp/fficli.cpp
@@ -239,6 +239,30 @@ cli_rnp_get_confirmation(const cli_rnp_t *rnp, const char *msg, ...)
     return false;
 }
 
+static bool
+rnp_ask_filename(const std::string &msg, std::string &res, cli_rnp_t &rnp)
+{
+    fprintf(rnp.userio_out, "%s", msg.c_str());
+    fflush(rnp.userio_out);
+    char        fname[128] = {0};
+    std::string path;
+    do {
+        if (!fgets(fname, sizeof(fname), rnp.userio_in)) {
+            return false;
+        }
+        path = path + std::string(fname);
+        if (rnp::strip_eol(path)) {
+            res = path;
+            return true;
+        }
+        if (path.size() >= 2048) {
+            fprintf(rnp.userio_out, "%s", "Too long filename, aborting.");
+            fflush(rnp.userio_out);
+            return false;
+        }
+    } while (1);
+}
+
 /** @brief checks whether file exists already and asks user for the new filename
  *  @param path output file name with path. May be an empty string, then user is asked for it.
  *  @param res resulting output path will be stored here.
@@ -249,41 +273,31 @@ cli_rnp_get_confirmation(const cli_rnp_t *rnp, const char *msg, ...)
 static bool
 rnp_get_output_filename(const std::string &path, std::string &res, cli_rnp_t &rnp)
 {
-    const size_t maxlen = PATH_MAX;
-    char         newpath[maxlen] = {0};
-    if (path.empty()) {
-        fprintf(rnp.userio_out, "Please enter the output filename: ");
-        fflush(rnp.userio_out);
-        if (!fgets(newpath, maxlen, rnp.userio_in)) {
-            return false;
-        }
-        rnp::strip_eol(newpath);
-    } else {
-        strncpy(newpath, path.c_str(), maxlen - 1);
-        newpath[maxlen - 1] = '\0';
+    std::string newpath = path;
+    if (newpath.empty() &&
+        !rnp_ask_filename("Please enter the output filename: ", newpath, rnp)) {
+        return false;
     }
 
     while (true) {
-        if (!rnp_file_exists(newpath)) {
+        if (!rnp_file_exists(newpath.c_str())) {
             res = newpath;
             return true;
         }
         if (rnp.cfg().get_bool(CFG_OVERWRITE) ||
             cli_rnp_get_confirmation(
-              &rnp, "File '%s' already exists. Would you like to overwrite it?", newpath)) {
-            rnp_unlink(newpath);
+              &rnp,
+              "File '%s' already exists. Would you like to overwrite it?",
+              newpath.c_str())) {
+            rnp_unlink(newpath.c_str());
             res = newpath;
             return true;
         }
 
-        fprintf(rnp.userio_out, "Please enter the new filename: ");
-        fflush(rnp.userio_out);
-        if (!fgets(newpath, maxlen, rnp.userio_in)) {
+        if (!rnp_ask_filename("Please enter the new filename: ", newpath, rnp)) {
             return false;
         }
-
-        rnp::strip_eol(newpath);
-        if (!strlen(newpath)) {
+        if (newpath.empty()) {
             return false;
         }
     }

--- a/src/tests/streams.cpp
+++ b/src/tests/streams.cpp
@@ -164,14 +164,14 @@ static void
 copy_tmp_path(char *buf, size_t buflen, pgp_dest_t *dst)
 {
     typedef struct pgp_dest_file_param_t {
-        int  fd;
-        int  errcode;
-        bool overwrite;
-        char path[PATH_MAX];
+        int         fd;
+        int         errcode;
+        bool        overwrite;
+        std::string path;
     } pgp_dest_file_param_t;
 
     pgp_dest_file_param_t *param = (pgp_dest_file_param_t *) dst->param;
-    strncpy(buf, param->path, buflen);
+    strncpy(buf, param->path.c_str(), buflen);
 }
 
 TEST_F(rnp_tests, test_stream_file)

--- a/src/tests/streams.cpp
+++ b/src/tests/streams.cpp
@@ -182,7 +182,7 @@ TEST_F(rnp_tests, test_stream_file)
     const char * filedata = "dummy message to be stored in the file";
     const int    iterations = 10000;
     const int    filedatalen = strlen(filedata);
-    char         tmpname[PATH_MAX] = {0};
+    char         tmpname[128] = {0};
     uint8_t      tmpbuf[1024] = {0};
     pgp_dest_t   dst = {};
     pgp_source_t src = {};

--- a/src/tests/support.cpp
+++ b/src/tests/support.cpp
@@ -206,17 +206,19 @@ get_tmp()
 static bool
 is_tmp_path(const char *path)
 {
-    char rlpath[PATH_MAX] = {0};
-    char rltmp[PATH_MAX] = {0};
-    if (!realpath(path, rlpath)) {
-        strncpy(rlpath, path, sizeof(rlpath));
-        rlpath[sizeof(rlpath) - 1] = '\0';
+    char *rlpath = realpath(path, NULL);
+    if (!rlpath) {
+        rlpath = strdup(path);
     }
     const char *tmp = get_tmp();
-    if (!realpath(tmp, rltmp)) {
-        strncpy(rltmp, tmp, sizeof(rltmp));
+    char *      rltmp = realpath(tmp, NULL);
+    if (!rltmp) {
+        rltmp = strdup(tmp);
     }
-    return strncmp(rlpath, rltmp, strlen(rltmp)) == 0;
+    bool res = rlpath && rltmp && !strncmp(rlpath, rltmp, strlen(rltmp));
+    free(rlpath);
+    free(rltmp);
+    return res;
 }
 
 /* Recursively remove a directory.

